Add groups for commands in help

aawsome opened the pull request:
Adds the possibility to group the commands in the help message.

closes #836 
closes #1271

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/spf13/cobra?pullRequest=1003) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/spf13/cobra?pullRequest=1003) before we can accept your contribution.<br/><hr/>**Alexander Weiss** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/spf13/cobra?pullRequest=1003) it.</sub>

jharshman commented:
@aawsome I think this is a good addition.  It's something I can put on the roadmap after 1.0.0 is released.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/spf13/cobra?pullRequest=1003) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/spf13/cobra?pullRequest=1003) before we can accept your contribution.<br/><hr/>**Alexander Weiss** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/spf13/cobra?pullRequest=1003) it.</sub>

mhewedy commented:
Any updates on this? or any workarounds exist?

aawsome commented:
I rebased this PR.
Now that 1.0.0 is out, is there anything I can do to get this PR merged?

itskingori commented:
Maybe we can nudge @jharshman for some visibility. I think should be able to provide some direction.

umarcor commented:
@aawsome would you mind adding some tests for this feature?

aawsome commented:
Added some tests. Almost all new code is covered by these tests. 

The exception is `AddCommandGroup` when 'helpCommand' is manually set.
(In this case most people would anyway directly specify the `Group` within the `Command` definition)

aawsome commented:
Is there anything I can do to help getting this PR merged?

nacx commented:
This has been open for a while and would be pretty nice to have this merged. Any chance anyone could have a look at it?

itskingori commented:
@aawsome Branch has conflicts. Could you rebase? So that if it's ready by the time @jharshman has a another look?

aawsome commented:
> @aawsome Branch has conflicts. Could you rebase? So that if it's ready by the time @jharshman has a another look?

is rebased.

umarcor commented:
@yuvalsosman, see #1496.

aawsome commented:
From my side, I'm fine with getting this merged. Actually realized that this PR is now waiting now for more than two years - that's quite some time...

Anyhow, if there is anything I can do, please tell me!

itskingori commented:
@jpmcb Any chance you could look at this?

RainbowMango commented:
@aawsome Seems we are ready to go, but you might have to solve the conflicts first.

aawsome commented:
I rebased and fixed the test that was failing bacause of the automatically-added `completion` command.

This PR should be now ready for merging.

ssoroka commented:
@marckhouzam @jpmcb Sorry to ping, but can we get this merged now?

umarcor commented:
@ssoroka see #1626.

aawsome commented:
Sorry if this sounds rude.

This PR is now open for 2 1/2 years. I always almost immediately answered to any question or request. Which was in last history mostly rebasing because of code changes in master. This PR was on the roadmap for 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.5.0 (or maybe not for all, but I'm not connected to release planing). Sorry, but this is really annoying!

I'm not working seriously with go for about one year now. This PR is still mergeable. @maintainers: Please decide if you like to merge this PR or close it if not. When the next request for rebasing comes I might simply close it instead of investing (most likely wasted) work into this PR.

fruitbutt commented:
@marckhouzam @jharshman adding to the next milestone has occurred numerous times over the 2.5 years this PR has been open. Many of us would like to see this merged so we can stop maintaining forks, and or ugly templating hacks.

jpmcb commented:
Hi all - I'm sorry (again) this PR has gotten stale and I understand your frustration; I really do. I've been trying now for years to get the different places I've worked at to support me and my work for this extremely important open source project.

But, long story short, it all comes down to time; I just have no time to support this project on my own and the companies I've worked for are not interested in allocating any of my time to cobra. But it often feels like if I don't, no one else will, and [I deeply fear what could happen if this project falls into disarray](https://www.youtube.com/watch?v=YBsDnXXW_d8). If you didn't know already, cobra is a core dependency in Kubernetes, Helm, Tanzu, Cluster API, Linkerd, Grafana, Istio, etc. etc. and sits at a really important intersection of the Go software supply chain. This is generally why my posture towards new changes is a hesitant one.

So, I give corba what little extra time I have in order to ensure we don't have [another faker.js situation on our hands.](https://news.ycombinator.com/item?id=25032105)

Please don't take your frustration out on the maintainers; we are doing our best with what few resources we have but we understand how frustrating this is for the Go community. If we should be frustrated at anyone, it's the endless number of companies that are more than happy to consume these open source libraries while not giving back.

---

I will make this a priority for the next release which I hope to cut within the next few weeks. Thanks all.

umarcor commented:
Maybe not introducing any relevant change for years in order to satisfy the needs of all those big companies gave the idea that they don't need to care about this project, because maintainers will take care of gatekeeping potential problems (even in development). I don't mean John's employers only, since this was happening far before he joined the project. Should the project evolve as an open source project which is not funded by any of those companies, maybe they would understand the need to do so.

jpmcb commented on user_guide.md:
```diff
@@ -447,6 +447,11 @@ command and flag definitions are needed.
+### Grouping commands in help
+
+Cobra supports grouping of available commands. Groups can either be explicitly defined by `AddGroup` and set by
+the `Group` element of a subcommand. If Groups are not explicitly defined they are implicitly defined.
```
> If Groups are not explicitly defined they are implicitly defined.

I think this is a bit confusing. Instead, it seems that the commands that don't have a group get bundled into the "Available commands" which is the current default behavior. When I originally read this, I wasn't sure what "implicitly defined" meant. Can we simplify this chunk?

aawsome replied:
I agree we maybe should remove the ability to implicitly define groups. I'll remove this passage here.

aawsome replied:
is changed.

jpmcb reviewed (approved):
Overall, this looks good. Tested this out with a large cobra based utility, added a few groups and it works well.

Just a small documentation nit and this will be good to merge. And if @aawsome doesn't have the bandwidth to address that, we can just get this in and look at it latter

Might also be good to get @marckhouzam's eyes on this since it's a big community ask!

marckhouzam commented on command.go:
```diff
@@ -31,6 +31,12 @@ import (
 
+// Structure to manage groups for commands
+type Group struct {
+	Group string
```
How about calling this field `ID` or something like that?  I find that `Group.Group` is confusing.

aawsome replied:
I'll change it to `Id`.

marckhouzam commented on command.go:
```diff
@@ -31,6 +31,12 @@ import (
+type Group struct {
+	Group string
+	Title string
+}
```
I assume we create this struct to allow users to create a group once with a title, and then use the `Group.Group` field as a short id to be used multiple times?  Was that the idea behind this struct?

jpmcb replied:
I'm also a bit confused on this. Can't we just match on a string instead having to allocate a new "group" structure in memory?

aawsome replied:
Yes, the idea was to use groups by an ID within the `Command` struct and have the title just set once. This also allows to rather easily change the title on just one place.

If you wish, I can remove the ID, but then we still need to save an array of Strings and call `AddGroup` for all groups to set the order of the groups.

marckhouzam replied:
I think the separation between ID and Title may be good.  Let's first keep it, reviewe the other changes, and then see how we feel about it.

marckhouzam commented on command.go:
```diff
@@ -1147,6 +1175,10 @@ func (c *Command) AddCommand(cmds ...*Command) {
+		// if Group is not defined generate a new one with same title
+		if x.Group != "" && !c.ContainsGroup(x.Group) {
+			c.AddGroup(&Group{Group: x.Group, Title: x.Group})
+		}
```
I wonder what is better for the user?  
Implicitly creating the group seems nice at first impression.  However, I wonder if it makes typos harder to notice?
Say I want a group of commands called "System:" (notice the colon), and for one of my commands I type `Group: "System"` with a missing colon.  This will implicitly create a new group.  It will be obvious looking at the help, but if this is a couple of levels of subcommands down, it may go unnoticed.

What we could do instead is `panic` if the group was not created in advance.  That way, the user will not miss such subtle errors.

The flip side of course is that if forces users to call `AddGroup` for each group.  Is that too heavy?

A benefit is that it forces users to specify a `Title` for the group, instead of sometimes using the id (`Group.Group`) as the title.

And a nice side-effect is that it would make the order of groups more obvious: groups will be displayed in the order they were created with `AddGroup()`

Any opinions on that?

jpmcb replied:
> What we could do instead is panic if the group was not created in advance. That way, the user will not miss such subtle errors.

I like the idea of not implicitly creating groups now that you bring it up but is panicing the right move here? Do we have a panic president elsewhere in cobra to surface user error?

> And a nice side-effect is that it would make the order of groups more obvious: groups will be displayed in the order they were created with AddGroup()

Yes, I think that's the obvious benefit. The unknown ordering of groups in the help messaging would be a poor user experience for people creating CLIs. My vote would be to be more explicit.

marckhouzam replied:
> I like the idea of not implicitly creating groups now that you bring it up but is panicing the right move here? 
> Do we have a panic president elsewhere in cobra to surface user error?

Two lines above 😉 
https://github.com/spf13/cobra/blob/212ea4078323771dc49b6f25a41d84efbaac3a4c/command.go#L1177-L1179
I rarely have original thoughts 🤣 

I have to admit that using `panic()` is something I used to think was exaggerated.   However, I now feel that as long as we know that the panic would be triggered immediately, then we know a developer would immediately realize their error and correct it.

aawsome replied:
I'll change that such that it panics. I agree that implicitly defining groups was maybe a bad idea.

aawsome replied:
is changed.

marckhouzam reviewed (commented):
This is really great!  Thanks @aawsome for your efforts and sorry it took so long to get reviewed.

I think it is worth putting just a little more effort to make sure the API we provide is good for users.  Once the API is released, we won't be able to change it anymore.

Here are a couple of thoughts: 
1. We will need a way to set the group for the default `completion` command. ~We could add a field to the existing `CompletionOptions` structure to set the Group~ Let's use the same technique as for the `help` command so that it is obvious for users.  So having a function `SetCompletionCommandGroup`.
1. If all commands are assigned to groups, we still see "Available Commands:" at the top but there are no commands under it.  Do we want that, or should we not show it in this case?  In truth, it is not that bad as it looks like it is a heading for all the groups.  However, I wonder if users will like that? (I've seen users refuse to upgrade to a Cobra version because the new `completion` command had its help start with a lower case instead of an uppercase 😄 )
2. If a command is not part of a group it ends up under "Available Commands:", with the Groups listed after.  I personally find "Available Commands:"  a poor title in this case as it will surely cause confusion compared to the group names.  How about we call the section "Additional commands:" and put it after all the groups? 
3. The ordering of the Groups in the help output is not obvious to a user, especially when using implicit creating of groups.  We would at least need to document it.  But let's see if we keep implicit creation or not (see in-line comments).
4. And a couple of questions inline.

marckhouzam reviewed (changes requested):
Marking that I request changes because we need at a minimum to allow to put the `completion` command in a group

jpmcb commented:
> If all commands are assigned to groups, we still see "Available Commands:" at the top but there are no commands under it. Do we want that, or should we not show it in this case?

Good point: we should not show "Available Commands" (or maybe something like Marc suggested "Additional commands") if there are no commands under it (aka, all commands have been added to groups. That would be the least confusing, best user experience.

aawsome commented:
>     1. We will need a way to set the group for the default `completion` command. ~We could add a field to the existing `CompletionOptions` structure to set the Group~ Let's use the same technique as for the `help` command so that it is obvious for users.  So having a function `SetCompletionCommandGroup`.

Will be added.

>     2. If all commands are assigned to groups, we still see "Available Commands:" at the top but there are no commands under it.  Do we want that, or should we not show it in this case?  In truth, it is not that bad as it looks like it is a heading for all the groups.  However, I wonder if users will like that? (I've seen users refuse to upgrade to a Cobra version because the new `completion` command had its help start with a lower case instead of an uppercase smile )

I thought to leave this as a header. It can be of course removed, but that would require a bit of computation to check if all commands have the Group set. Should I do this?

>     3. If a command is not part of a group it ends up under "Available Commands:", with the Groups listed after.  I personally find "Available Commands:"  a poor title in this case as it will surely cause confusion compared to the group names.  How about we call the section "Additional commands:" and put it after all the groups?

Ok, I can also add this, but I think we should keep the exiting behavior if the user defines no group at all. That is, if no group is defined, the "group title" of a command without a `Group` set is "Available Commands" and if at least one group is defined, it would be "Additional Commands".
Also, if users wants to see "Additional Commands", they can also get this behavior by defining a new group titled "Additional Commands" and put these commands into that group.
So, do we really want this?

>     4. The ordering of the Groups in the help output is not obvious to a user, especially when implicitly creating groups.  We would at least need to document it.  But let's see if we keep implicit creation or not (see in-line comments).

After removing the implicit definition, the order will be the order of the `AddGroup` commands. I'll add this in the manual.

marckhouzam commented:
> > ```
> > 2. If all commands are assigned to groups, we still see "Available Commands:" at the top but there are no commands under it.  Do we want that, or should we not show it in this case?  In truth, it is not that bad as it looks like it is a heading for all the groups.  However, I wonder if users will like that? (I've seen users refuse to upgrade to a Cobra version because the new `completion` command had its help start with a lower case instead of an uppercase smile )
> > ```
> 
> 
> I thought to leave this as a header. It can be of course removed, but that would require a bit of computation to check if all commands have the Group set. Should I do this?

Leave it for now and we'll see how we feel on the next review.

> 
> > ```
> > 3. If a command is not part of a group it ends up under "Available Commands:", with the Groups listed after.  I personally find "Available Commands:"  a poor title in this case as it will surely cause confusion compared to the group names.  How about we call the section "Additional commands:" and put it after all the groups?
> > ```
> 
> 
> Ok, I can also add this, but I think we should keep the exiting behavior if the user defines no group at all. That is, if no group is defined, the "group title" of a command without a `Group` set is "Available Commands" and if at least one group is defined, it would be "Additional Commands".

Good point. This is actually important to keep the help the same for existing programs 👍

> Also, if users wants to see "Additional Commands", they can also get this behavior by defining a new group titled "Additional Commands" and put these commands into that group.
> So, do we really want this?

The "Additional Commands" automatic group is just to catch commands that were not put into any group.  It should not actually happen but experience shows we better play it safe.  It may happen fire the `help` and `completion` command where devs may forget to give them a group or not realize they can.

If it's complicated to implement, leave it and we can see on the next review.

> 
> 
> > ```
> > 4. The ordering of the Groups in the help output is not obvious to a user, especially when implicitly creating groups.  We would at least need to document it.  But let's see if we keep implicit creation or not (see in-line comments).
> > ```
> 
> 
> After removing the implicit definition, the order will be the order of the `AddGroup` commands. I'll add this in the manual.

Perfect.

Looking forward to the next iteration! 🎉

aawsome commented:
I rebased this PR to the current master and made the mentioned changes.
I now get a linting error `Error: struct of size 720 bytes could be of size 712 bytes (maligned)` which I don't know how to solve. And IMO I will add a test of `SetCompletionCommandGroup`.

When these two points (and squashing the commits) are tackled, I think this PR  will be ready.

marckhouzam commented on command.go:
```diff
@@ -1200,6 +1241,26 @@ func (c *Command) AddCommand(cmds ...*Command) {
+	return c.commandgroups
+}
+
+// ContainGroups return if group is in command groups.
```
```suggestion
// ContainGroups return if groupId exists in the list of command groups.
```

marckhouzam commented on command.go:
```diff
@@ -1200,6 +1241,26 @@ func (c *Command) AddCommand(cmds ...*Command) {
+}
+
+// ContainGroups return if group is in command groups.
+func (c *Command) ContainsGroup(group string) bool {
```
```suggestion
func (c *Command) ContainsGroup(groupId string) bool {
```

marckhouzam reviewed (commented):
Some small suggestions to cleanup a bit.

I haven't test yet.

marckhouzam commented:
> I rebased this PR to the current master and made the mentioned changes. 
> I now get a linting error `Error: struct of size 720 bytes could be of size 712 bytes (maligned)` which I don't know how to solve.

If you move the new `completionCommandGroup` declaration slightly higher right under the declaration of `helpCommandGroup`, the linter error goes away.  It has to do with memory alignment, although I haven't looked exactly at what was not well aligned.

> And IMO I will add a test of `SetCompletionCommandGroup`.

Yes please.

marckhouzam reviewed (changes requested):
Ok, I like this a lot.  Nice job @aawsome.  I put a bunch of minor suggestions about improving comments and such, but you should just be able to accept them in Github which should be very easy.

I believe the only thing remaining is when a command is not part of a group.  I have created the "Administration Commands:" and "System Commands:" groups but "forgot" to assign the `login` command to any group.
Currently we see this:
```
My CLI

Usage:
  prog [command]

Available Commands:
  login             Login to the platform

System Commands:
  config            Configuration for the CLI
  init              Initialize the CLI
  update            Update the CLI
  version           Version information

Administration Commands:
  completion        Generate the autocompletion script for the specified shell
  help              Help about any command

Flags:
  -h, --help   help for prog
```

I think the section "Available Commands:" looks bad.
However, I would personally be ok with leaving it as is and saying that it is up to developers to properly define their groups and not leave commands without a group.

@jpmcb Thoughts?

@aawsome I'm finished with the review.  Whenever you find time, if you can update, and then we'll be pretty much ready to merge 🎉 .

aawsome commented on command.go:
```diff
@@ -1178,6 +1215,10 @@ func (c *Command) AddCommand(cmds ...*Command) {
 		cmds[i].parent = c
+		// if Group is not defined generate a new one with same title
+		if x.Group != "" && !c.ContainsGroup(x.Group) {
+			panic("Group is not defined")
```
is added.

aawsome commented on command.go:
```diff
@@ -61,6 +67,9 @@ type Command struct {
 	Short string
 
+	// The group under which the command is grouped in the 'help' output.
+	Group string
```
The linter says `GroupID`, so I used that. I also added the `ID` at all places where the id is actually used, e.g. `SetHelpCommandGroupID` etc.

aawsome commented:
Thanks a lot for your review, @marckhouzam 
I resolved all comments, added a test for the `completion` command and rebased this PR.
I hope now everything is fine :wink:

jpmcb reviewed (commented):
> I would personally be ok with leaving it as is and saying that it is up to developers to properly define their groups and not leave commands without a group.

I think this is should be the desired behavior since it essentially makes this feature backwards compatible: if someone doesn't define _any_ groups, I'd expect this the help section to continue to work like it does today.

---

I've pondered the design decisions this PR carries and while I think this implementation is good, I wonder if this would be due for a refactor down the road since people may want a way to do more interesting things with groups beyond matching on strings (i.e., a group struct with a superset of cobra commands). But we'd need to see if people would even want that. 

TLDR: I think this is good for now and we can merge it to get this behavior in

marckhouzam commented:
I tested the latest version and it looked great.
But I kept being bothered by the "Additional commands:" section, mostly when it was empty.
So I went ahead and pushed a new commit that fixed it.  This is the result.

If no groups are defined at all we remain backwards-compatible and get:
```
$ prog -h
My CLI

Usage:
  prog [command]

Available Commands:
  completion        Generate the autocompletion script for the specified shell
  config            Configuration for the CLI
  help              Help about any command
  version           Version information

Flags:
  -h, --help   help for prog

Use "prog [command] --help" for more information about a command.
```
If all subcommands have a group, we no longer get an empty "Available Commands:" line:
```
$ prog -h
My CLI

Usage:
  prog [command]

System
  config            Configuration for the CLI
  version           Version information

Administration
  completion        Generate the autocompletion script for the specified shell
  help              Help about any command

Flags:
  -h, --help   help for prog

Use "prog [command] --help" for more information about a command.
```
And if some commands have groups but some do not we get a new "Additional Commands:" section at the bottom:
```
$ prog -h
My CLI

Usage:
  prog [command]

System
  config            Configuration for the CLI

Administration
  version           Version information

Additional Commands:
  completion        Generate the autocompletion script for the specified shell
  help              Help about any command

Flags:
  -h, --help   help for prog

Use "prog [command] --help" for more information about a command.
```

@aawsome @jpmcb do you agree this is better?  If not, we can just remove the top commit that I pushed.
If you like it, I'm ready to merge.

> I wonder if this would be due for a refactor down the road since people may want a way to do more
> interesting things with groups beyond matching on strings (i.e., a group struct with a superset of cobra commands).
>  But we'd need to see if people would even want that.

That is interesting and may indeed have potential for improvement.  But I also can't tell at the moment what that would look like and if it will be in demand.  So I agree with @jpmcb that we should go ahead with this version, which we know people want, and see where it takes us.

aawsome commented:
> So I went ahead and pushed a new commit that fixed it.

Just had a look at your commit. This is how I would have added these functionalities. So,your changes look good for me!

jpmcb reviewed (approved):
Amazing work @marckhouzam!!

Airblader commented:
I'd just like to say thank you from a backseat watcher and user of the library to @aawsome for bringing this over the finish line even after such a long time! And of course everyone else who helped along the way.

marckhouzam commented:
Yes thank you @aawsome !  You have shown great professionalism in what can be considered a frustrating situation 🙂

The community will reap the rewards 🎉

Airblader commented:
If you don't mind, a follow-up question while adopting this: the groups appear in the order they are defined, but what about the commands in the group? Can we order them? It seems that currently this is alphabetical, which isn't usually very meaningful.
